### PR TITLE
Multi-user Downlink Improvement

### PR DESCRIPTION
**Pull request by rdoosty**
- Improvement in calibration averaging
- Some bugfixes in client code

### CODE REVIEW

**Review by asedlmayr — APPROVED** [review]
Everything looks good to me

File: src/agora/dozf.cc [hunk 1]
```diff
@@ -223,34 +242,105 @@ void DoZF::ZfTimeOrthogonal(size_t tag) {
...
+                           "online_calib_vec_inst" +
+                               std::to_string(frame_id / cfg_->AntGroupNum()),
+                           true /*append*/);
+            Utils::SaveVec(calib_vec, "calib_vec.m",
+                           "online_calib_vec_mean" +
+                               std::to_string(frame_id / cfg_->AntGroupNum()),
+                           true /*append*/);
+          }
+        }
+      }
```
Comment (jianding17) [inline src/agora/dozf.cc:343]: Can we modularize this part of code (separating it into smaller functions)? This function too long now and this part of code might need to be applied to frequency orthogonal case in the future.

File: src/agora/dozf.cc [hunk 2]
```diff
@@ -103,6 +120,8 @@ void DoZF::ComputePrecoder(const arma::cx_fmat& mat_csi,
         arma::cx_fmat(cfg_->UeNum(), cfg_->NumChannels(), arma::fill::zeros));
   }
   mat_ul_zf = mat_ul_zf_tmp;
+  float rcond = arma::rcond(mat_csi.t() * mat_csi);
+  return rcond;
```
Comment (jianding17) [inline src/agora/dozf.cc:124]: The computation of the condition number seems to not cheap. It seems this number is only used in PhyStats. Can we only enable the computation when necessary?

File: src/agora/dozf.cc [hunk 3]
```diff
@@ -261,9 +351,10 @@ void DoZF::ZfTimeOrthogonal(size_t tag) {
     double start_tsc3 = GetTime::WorkerRdtsc();
     duration_stat_->task_duration_[2] += start_tsc3 - start_tsc2;
 
-    ComputePrecoder(mat_csi, calib_gather_buffer_,
-                    ul_zf_matrices_[frame_slot][cur_sc_id],
-                    dl_zf_matrices_[frame_slot][cur_sc_id]);
+    auto rcond = ComputePrecoder(mat_csi, calib_gather_buffer_,
+                                 ul_zf_matrices_[frame_slot][cur_sc_id],
+                                 dl_zf_matrices_[frame_slot][cur_sc_id]);
+    phy_stats_->UpdateCsiCond(frame_id, cur_sc_id, rcond);
```
Comment (jianding17) [inline src/agora/dozf.cc:357]: We should add a flag to allow disabling phy_stats update for performance test.

File: src/agora/txrx/txrx.cc [hunk 4]
```diff
@@ -183,6 +194,8 @@ void PacketTXRX::LoopTxRx(size_t tid) {
 
   MLPD_INFO("LoopTxRx[%zu] has %zu:%zu total radios %zu\n", tid, radio_lo,
             radio_hi, (radio_hi - radio_lo) + 1);
+  //\todo Sync the start of the threads
```
Comment (jianding17) [inline src/agora/txrx/txrx.cc:197]: The comments need to have space between // and text.

**Review by rdoosty — COMMENTED** [review]
I have addressed the comments @jianding17. Please review again. Thanks!

**Pull request merged**